File: remote/src/main/java/com/kelsos/mbrc/presenters/PlaylistTrackPresenterImpl.java
package com.kelsos.mbrc.presenters;

import com.google.inject.Inject;
import com.kelsos.mbrc.annotations.Queue;
import com.kelsos.mbrc.domain.PlaylistTrack;
import com.kelsos.mbrc.interactors.QueueInteractor;
import com.kelsos.mbrc.interactors.playlists.PlaylistTrackInteractor;
import com.kelsos.mbrc.ui.views.PlaylistTrackView;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import timber.log.Timber;

public class PlaylistTrackPresenterImpl implements PlaylistTrackPresenter {
  private PlaylistTrackView view;

  @Inject private PlaylistTrackInteractor playlistTrackInteractor;
  @Inject private QueueInteractor queueInteractor;

  @Override public void bind(PlaylistTrackView view) {
    this.view = view;
  }

  @Override public void load(long playlistId) {
    playlistTrackInteractor.execute(playlistId)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(playlistTracks -> {
          view.update(playlistTracks);
        }, throwable -> {
          view.showErrorWhileLoading();
          Timber.e(throwable, "");
        });
  }

  @Override public void queue(PlaylistTrack track, @Queue.Action String action) {
    queueInteractor.execute(action, track.getPath()).subscribe(aBoolean -> {

    }, throwable -> {
      Timber.e(throwable, "Queueing failed");
    });
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/presenters/PlaylistTrackPresenterImpl.kt
package com.kelsos.mbrc.presenters

import javax.inject.Inject
import com.kelsos.mbrc.annotations.Queue
import com.kelsos.mbrc.domain.PlaylistTrack
import com.kelsos.mbrc.extensions.task
import com.kelsos.mbrc.interactors.QueueInteractor
import com.kelsos.mbrc.interactors.playlists.PlaylistTrackInteractor
import com.kelsos.mbrc.ui.views.PlaylistTrackView
import timber.log.Timber

class PlaylistTrackPresenterImpl : PlaylistTrackPresenter {
  private var view: PlaylistTrackView? = null

  @Inject private lateinit var playlistTrackInteractor: PlaylistTrackInteractor
  @Inject private lateinit var queueInteractor: QueueInteractor

  override fun bind(view: PlaylistTrackView) {
    this.view = view
  }

  override fun load(longExtra: Long) {
    playlistTrackInteractor.execute(longExtra)
        .task()
        .subscribe({ view?.update(it) },
            {
              view?.showErrorWhileLoading()
              Timber.e(it, "")
            })
  }

  override fun queue(track: PlaylistTrack, @Queue.Action action: String) {
    if (track.path.isNullOrEmpty()) {
      return
    }

    queueInteractor.execute(action, track.path).subscribe({

    }) { Timber.e(it, "Queueing failed") }
  }
}
